TYPO3.Media/Classes/TYPO3/Media/Domain/Service/ThumbnailService.php

```diff
@@ -86,6 +86,19 @@ class ThumbnailService
      */
     public function getThumbnail(AssetInterface $asset, ThumbnailConfiguration $configuration)
     {
+        // Calculates the dimensions of the thumbnail to be generated and returns the thumbnail image if the new
+        // dimensions differ from the specified image dimensions, otherwise the original image is returned.
+        if ($asset instanceof ImageInterface) {
+            if ($asset->getWidth() === null && $asset->getHeight() === null) {
+                return $asset;
+            }
+            $maximumWidth = ($configuration->getMaximumWidth() > $asset->getWidth()) ? $asset->getWidth() : $configuration->getMaximumWidth();
+            $maximumHeight = ($configuration->getMaximumHeight() > $asset->getHeight()) ? $asset->getHeight() : $configuration->getMaximumHeight();
+            if ($configuration->isUpScalingAllowed() === false && $maximumWidth === $asset->getWidth() && $maximumHeight === $asset->getHeight()) {
+                return $asset;
```

Comment:
Here and in the return statement above an Asset is returned, but the return value is documented to be a thumbnail. That has been mentioned by @dfeyer already and should be fixed.